Databricks external location setting (#606)
rudderlabs/rudderstack-docs: docs/data-warehouse-integrations/delta-lake.mdx

For more information on generating the access token, refer to the <a href="https://rudderstack.com/docs/data-warehouse-integrations/delta-lake/#generating-the-databricks-access-token">Generating the Databricks access token</a> section below.
</div>

> * **Enable delta tables creation in an external location**: Disabled by default, Rudderstack creates delta tables at a default storage location for non-external Hive tables.

Reviewer: ```suggestion
* **Enable delta tables creation in an external location**: RudderStack creates the delta tables at a default storage location for the non-external Apache Hive tables. By default, this setting is disabled. 
```

  When enabled, it lets you specify the external location to create delta tables: 

>     <img  src="../assets/dw-integrations/delta-lake-connection-settings-13.png" alt="Delta Lake destination settings in RudderStack"/>

Reviewer: ```suggestion
<img  src="../assets/dw-integrations/delta-lake-connection-settings-13.png" alt="Delta Lake destination settings in RudderStack"/>
```

>     The delta tables can further be accessed at `{externalLocation}/{schema}/{table}`.

Reviewer: ```suggestion
<div class="infoBlock">

The delta tables can be accessed at <code class="inline-code">{externalLocation}/{schema}/{table}</code>.
</div>
```

>   <div class="infoBlock">

Reviewer: ```suggestion
  <div class="warningBlock">
```

>    Note that creating delta tables in an external location is a beta feature currently.<img src="https://img.shields.io/badge/Stability-Beta-blueviolet" alt="Github Badge" class="githubBadge" />

Reviewer: Should we simply say something like "This feature is currently in beta"? Also, I think this should be a warning block for more emphasis on using it carefully with the production data. Also, I don't think we need to include the beta badge here - adding this note should be enough IMO.

Also, integers are **not** allowed at the start of a schema or table name. Hence, such schema, column, or table names will be prefixed with a `_`. For example, `'25dollarpurchase'` will be changed to `'_25dollarpurchase'`.

### How can I modify an existing table to a partitioned table?

> 1. Set an exclusion window (using the **Exclude window** connection setting) so that RudderStack does not process anything while performing the below changes.

Reviewer: ```suggestion
To modify an existing table to a partitioned table, follow these steps:

1. Set an exclusion window (using the **Exclude window** connection setting) so that RudderStack does not process anything while performing the below changes.
```

> 2. Make the required changes in Rudderstack Warehouse configurations.

Reviewer: just to confirm, "RudderStack warehouse" means the delta lake destination here, right? cc @achettyiitr

Reviewer 2: I mean this.
<img width="1792" alt="Screenshot 2022-05-24 at 5 21 10 PM" src="https://user-images.githubusercontent.com/77991816/170028199-42734263-2986-424a-9f85-800b556b87e0.png">
.

Reviewer: got it!

3. Run the following queries in the Databricks Cluster/SQL Endpoints to:
    - Rename the existing table with `_temp` suffix.
    - Add `event_date` column to the `_temp` table.
    - Backfill the data into original table.

```sql
ALTER TABLE x RENAME TO x_temp;
ALTER TABLE x_temp ADD COLUMN TO event_date DATE;
INSERT INTO x SELECT * FROM x_temp;
```

<div class="infoBlock">

The new tables will be created with partition support. Refer to the Databricks document for more information on [Partitions](https://docs.databricks.com/sql/language-manual/sql-ref-partition.html).

</div>

> ### How do I convert an existing unmanaged table at a location to an unmanaged table at a new location? / How do I convert an existing managed table to an unmanaged table at a new location?

Reviewer: Can we combine and rephrase both these questions to:

"How can I convert an existing table at a particular location to an unmanaged table at a new location?" cc @ishachopra0401 @achettyiitr

Author: IMO, a user might have any of these cases and being explicit in the ques about "existing **unmanaged** table at a location" and "an existing **managed** table to..." can actually help.

Reviewer: got it - in that case - "How can I convert an existing managed or unmanaged table at a location to an unmanaged table at a new location?" would be a good alternative IMO.

Reviewer 2: > How can I convert an existing managed or unmanaged table at a location to an unmanaged table at a new location?
This looks clean.

> 1. Set an exclusion window (using the **Exclude window** connection setting) so that RudderStack does not process anything while performing the below changes.

Reviewer: ```suggestion
1. Set an exclusion window (using the **Exclude window** connection setting) so that RudderStack does not process any data while performing the below changes.
```

> 2. Run the following queries in the Databricks Cluster/SQL Endpoints to:

Reviewer: ```suggestion
2. Run the following queries in the Databricks Cluster/SQL endpoints to:
```

>    - Create a temporary table using new location.

Reviewer: ```suggestion
   - Create a temporary table using the new location.
```

>    - Drop temporary table.

Reviewer: ```suggestion
   - Drop the temporary table.
```

>    - Drop original table.

Reviewer: ```suggestion
   - Drop the original table.
```

```sql
CREATE OR REPLACE TABLE namespace.x_temp DEEP CLONE namespace.x LOCATION '/path/to/new/location/namespace/x';
// where namespace represents the namespace attached to the destination in RudderStack.
// where x represents the original table created by RudderStack.

DROP TABLE namespace.x_temp;
DROP TABLE namespace.x;
```

> 3. Enable the **Enable delta tables creation in an external location** setting in Rudderstack dashboard and update the location.

Reviewer: ```suggestion
3. Enable the **Enable delta tables creation in an external location** setting in RudderStack dashboard and update the location.
```

> 4. Remove the exclusion window and make the required changes in Rudderstack Warehouse configurations.

Reviewer: Same query as above - this refers to the delta lake destination configuration in the dashboard, right?

Reviewer 2: No.

Reviewer 2: Same as [this](https://github.com/rudderlabs/rudderstack-docs/pull/606#discussion_r880406823).

> The table will be created again during the subsequent data syncs.

Reviewer: ```suggestion
RudderStack will create the table again during the subsequent data syncs.
```

<div class="infoBlock">

> Refer to the Databricks document for more information on [Managed and unmanaged tables](https://docs.databricks.com/data/tables.html#managed-and-unmanaged-tables).

Reviewer: ```suggestion
Refer to the [Databricks documentation on managed and unmanaged tables](https://docs.databricks.com/data/tables.html#managed-and-unmanaged-tables) for more information.
```

</div>

### How do I convert an existing unmanaged table at a specific location to a managed table?

> 1. Set an exclusion window (using the **Exclude window** connection setting) so that RudderStack does not process anything while performing the below changes.

Reviewer: ```suggestion
To convert an existing unmanaged table at a given location to a managed table, follow these steps:

1. Set an exclusion window (using the **Exclude window** connection setting) so that RudderStack does not process anything while performing the below changes.
```

Reviewer: also, is the managed table created at the same location? cc @achettyiitr @ishachopra0401

Author: AFAIU, yes.

Reviewer: we should explicitly mention that in the lead-in sentence, in that case, for clarity.

Reviewer 2: No, managed table are created at default location set in the Databricks.

2. Run the following queries in the Databricks Cluster/SQL Endpoints to:
    - Create a temporary table.
    - Drop original table.
    - Rename temporary table to original table.

```sql
CREATE TABLE IF NOT EXISTS namespace.x_temp DEEP CLONE namespace.x;
// where namespace represents the namespace attached to the destination in RudderStack.
// where x represents the original table created by RudderStack.

DROP TABLE namespace.x;
ALTER TABLE namespace.x_temp RENAME TO namespace.x;
```

> 3. Remove exclusion window and make sure the **Enable delta tables creation in an external location** setting is disabled in the Rudderstack dashboard.

Reviewer: ```suggestion
3. Remove the exclusion window and make sure the **Enable delta tables creation in an external location** setting is disabled in the RudderStack dashboard.
```

> The table will be created again during the subsequent data syncs.

Reviewer: ```suggestion
RudderStack will create the table again during the subsequent data syncs.
```
